File: src/pymergetic/metal/mem/.pm/smoke.c
/* Host C ABI smoke — metal mod test (links host libmetal_mem). */
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "__init__.h"

#define ARENA_BYTES (512u * 1024u)
#define ARENA_ALIGN 4096u

int main(void)
{
  uint8_t *raw;
  uint8_t *base;
  uint8_t *a;
  uint8_t *b;
  uint8_t *p0;
  uint8_t *p1;
  uint8_t *r;
  int i;
  int n_ok;

  /*
   * Host-only seed buffer (fake). On iron this range comes from
   * BIOS/UEFI/DT RAM — not from pm_metal_mem_alloc (needs init first).
   * Plain malloc + align — no C11/POSIX feature macros (clangd-safe).
   */
  raw = (uint8_t *)malloc(ARENA_BYTES + ARENA_ALIGN);
  if (raw == NULL) {
    fprintf(stderr, "mem .pm/smoke.c: malloc failed\n");
    return 1;
  }
  base = (uint8_t *)(((uintptr_t)raw + (ARENA_ALIGN - 1u)) & ~(uintptr_t)(ARENA_ALIGN - 1u));
  if (pm_metal_mem_init(base, ARENA_BYTES) != 0) {
    fprintf(stderr, "mem .pm/smoke.c: init failed\n");
    free(raw);
    return 1;
  }

  a = pm_metal_mem_alloc(64);
  b = pm_metal_mem_alloc(128);
  if (a == NULL || b == NULL) {
    fprintf(stderr, "mem .pm/smoke.c: alloc failed\n");
    return 1;
  }
  pm_metal_mem_free(b);
  pm_metal_mem_free(a);

  r = pm_metal_mem_alloc(32);
  if (r == NULL) {
    return 1;
  }
  memset(r, 0xA5, 32);
  r = pm_metal_mem_realloc(r, 256);
  if (r == NULL || r[0] != (uint8_t)0xA5) {
    fprintf(stderr, "mem .pm/smoke.c: realloc failed\n");
    return 1;
  }
  r = pm_metal_mem_realloc(r, 0);
  if (r != NULL) {
    return 1;
  }

  r = pm_metal_mem_memalign(64, 100);
  if (r == NULL || ((uintptr_t)r % 64u) != 0) {
    fprintf(stderr, "mem .pm/smoke.c: memalign failed\n");
    return 1;
  }
  pm_metal_mem_free(r);
  if (pm_metal_mem_memalign(3, 16) != NULL) {
    fprintf(stderr, "mem .pm/smoke.c: memalign bad align accepted\n");
    return 1;
  }

  p0 = pm_metal_mem_map(4096);
  p1 = pm_metal_mem_map(4096);
  if (p0 == NULL || p1 == NULL) {
    return 1;
  }
  if (pm_metal_mem_unmap(p0, 4096) != -1) {
    fprintf(stderr, "mem .pm/smoke.c: LIFO expected fail\n");
    return 1;
  }
  if (pm_metal_mem_unmap(p1, 4096) != 0 || pm_metal_mem_unmap(p0, 4096) != 0) {
    fprintf(stderr, "mem .pm/smoke.c: LIFO unmap failed\n");
    return 1;
  }

  n_ok = 0;
  for (i = 0; i < 64; i++) {
    a = pm_metal_mem_alloc(8 * 1024);
    if (a == NULL) {
      break;
    }
    n_ok++;
  }
  if (n_ok < 4) {
    fprintf(stderr, "mem .pm/smoke.c: pressure too weak (%d)\n", n_ok);
    return 1;
  }

  free(raw);
  printf("mem .pm/smoke.c: PASS\n");
  return 0;
}

File: src/pymergetic/metal/mem/.pm/smoke.rs
//! Host smoke — ``metal mod test mem`` / ``cargo run --bin smoke``.
use std::alloc::{alloc, dealloc, Layout};
use std::sync::atomic::{AtomicUsize, Ordering};
use std::thread;

use pymergetic_metal_mem::lock::{Mutex, MutexCell, Spin};

static SPIN: Spin = Spin::new();
static SPIN_COUNT: AtomicUsize = AtomicUsize::new(0);

fn main() {
    // --- spin / mutex (uncontended) ---
    let s = Spin::new();
    assert!(s.try_lock());
    assert!(!s.try_lock());
    s.unlock();
    s.lock();
    s.unlock();

    let m = Mutex::new();
    assert!(m.try_lock());
    assert!(!m.try_lock());
    m.unlock();
    m.lock();
    m.unlock();

    let cell = MutexCell::new(7u32);
    {
        let mut g = cell.lock();
        assert_eq!(*g, 7);
        *g = 9;
    }
    assert_eq!(*cell.lock(), 9);

    // --- spin contention (multi-thread) ---
    SPIN_COUNT.store(0, Ordering::SeqCst);
    let mut handles = Vec::new();
    for _ in 0..4 {
        handles.push(thread::spawn(|| {
            for _ in 0..2500 {
                SPIN.lock();
                let v = SPIN_COUNT.load(Ordering::Relaxed);
                SPIN_COUNT.store(v + 1, Ordering::Relaxed);
                SPIN.unlock();
            }
        }));
    }
    for h in handles {
        h.join().unwrap();
    }
    assert_eq!(SPIN_COUNT.load(Ordering::SeqCst), 10_000);

    // Smaller arena so TLSF seed exhausts and grow_and_add_pool runs.
    // Host-only seed buffer (fake). On iron this range comes from
    // BIOS/UEFI/DT RAM — not from pymergetic_metal_mem::api::alloc (needs init first).
    const N: usize = 512 * 1024;
    let layout = Layout::from_size_align(N, 4096).unwrap();
    let base = unsafe { alloc(layout) };
    assert!(!base.is_null());

    unsafe {
        assert_eq!(pymergetic_metal_mem::api::init(base, N), 0);

        // --- malloc / free ---
        let a = pymergetic_metal_mem::api::alloc(64);
        let b = pymergetic_metal_mem::api::alloc(128);
        assert!(!a.is_null() && !b.is_null());
        pymergetic_metal_mem::api::free(b);
        pymergetic_metal_mem::api::free(a);

        // --- realloc (grow in place or move+copy) ---
        let r = pymergetic_metal_mem::api::alloc(32);
        assert!(!r.is_null());
        r.write_bytes(0xA5, 32);
        let r2 = pymergetic_metal_mem::api::realloc(r, 256);
        assert!(!r2.is_null());
        assert_eq!(*r2, 0xA5);
        let r3 = pymergetic_metal_mem::api::realloc(r2, 0);
        assert!(r3.is_null());

        // --- memalign ---
        let al = pymergetic_metal_mem::api::memalign(64, 100);
        assert!(!al.is_null());
        assert_eq!(al as usize % 64, 0);
        pymergetic_metal_mem::api::free(al);
        assert!(pymergetic_metal_mem::api::memalign(3, 16).is_null()); // not power of two

        // --- map LIFO ---
        let p0 = pymergetic_metal_mem::api::map(4096);
        let p1 = pymergetic_metal_mem::api::map(4096);
        assert!(!p0.is_null() && !p1.is_null());
        assert_eq!(p0 as usize % 4096, 0);
        assert_eq!(pymergetic_metal_mem::api::unmap(p0, 4096), -1); // not top
        assert_eq!(pymergetic_metal_mem::api::unmap(p1, 4096), 0);
        assert_eq!(pymergetic_metal_mem::api::unmap(p0, 4096), 0);

        // --- pressure / grow-on-OOM ---
        let mut n_ok = 0usize;
        let mut last = core::ptr::null_mut();
        for _ in 0..64 {
            let p = pymergetic_metal_mem::api::alloc(8 * 1024);
            if p.is_null() {
                break;
            }
            last = p;
            n_ok += 1;
        }
        assert!(n_ok >= 4, "expected several allocs before arena empty, got {n_ok}");
        // Still able to free last and reclaim
        if !last.is_null() {
            pymergetic_metal_mem::api::free(last);
            let again = pymergetic_metal_mem::api::alloc(8 * 1024);
            assert!(!again.is_null());
            pymergetic_metal_mem::api::free(again);
        }

        dealloc(base, layout);
    }
    println!("mem .pm/smoke.rs: PASS");
}
